### ISSUE TITLE
ExoPlayer's ConcatenatingMediaSource misinteracts with seekToNext()

### ISSUE THREAD

**Issue by stoyicker** [androidx/media#640]
### Version

Media3 1.1.1

### Devices that reproduce the issue

Google Pixel 4a

### Devices that do not reproduce the issue

None found

### Reproducible in the demo app?

No

### Reproduction steps

1. Check out https://github.com/stoyicker/mwe-concatenatingmediasource/tree/43c81cf2ff12b6cdf6a93bbaa000a0fd09515bf1
2. Run the app

### Expected result

1. The item with the URI finishing in ranum_24bits.flac starts playback
2. The item with the URI finishing in tears_audio_eng.mp4 starts playback

(1 likely lasts too little to be noticeable, but at least it is reasonable to expect 2)

### Actual result

Playback never starts as corroborated by log calls from the app pertaining playback state changes.

Additionally, I'd like to note that posting these lines:

```
setMediaSource(concatenatingMediaSource)
playWhenReady = true
prepare()
```
after the dispatch of the first media source addition causes the first item to be played, and then when it finishes playback of the third item starts, which gives me the impression that the `seekToNext()` called skipped not the currently playing item when it was received, but the next one instead. Neither this behavior nor the one from the MWE as is (with these lines before dispatching any addMediaSource calls on the ConcatenatingMediaSource) seem correct to me.

### Media

All required media is already available in the linked MWE.

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to android-media-github@google.com after filing this issue.

**Comment by marcbaechinger** [comment]
When I try the app and add an `EventLogger` then I see that the player is prepared when the `ConcatenatingMediaSource` is still empty. The player then falls into `STATE_ENDED` and afterwards media sources are added and removed.

With this setup I think this is working as intended. See logs below.

I think this is a threading problem created by the sample app and the calls to ExoPlayer do not have the expected sequence. Mostly, prepare is called before media sources are ended to the concatenating media source.

I think if the calls to ExoPlayer are done in the correct sequence this just works.

Can you try doing the setup without special threads as a start? Then if the app really needs these additional threads, can you make sure that media sources are added before `prepare()` is called on the player?

```

ExoPlayerImpl       I  Init 70c5693 [AndroidXMedia3/1.1.1] [raven, Pixel 6 Pro, Google, 33]
mwe.root            W  Accessing hidden method Landroid/media/AudioTrack;->getLatency()I (unsupported, reflection, allowed)
EventLogger         D  playWhenReady [eventTime=0.00, mediaPos=0.00, window=0, true, USER_REQUEST]
EventLogger         D  state [eventTime=0.00, mediaPos=0.00, window=0, ENDED]
TEST-Thread[main]   D  Finished dispatching all commands
Compatib...eporter  D  Compat change id reported: 210923482; UID 10254; state: ENABLED
TEST-Thr...Thread]  D  Finished addMediaSource for item id=mp4 on index 0
Compatib...eporter  D  Compat change id reported: 237531167; UID 10254; state: DISABLED
TEST-Thr...Thread]  D  Finished addMediaSource for item id=mp4 on index 1
TEST-Thr...Thread]  D  Finished addMediaSource for item id=mp4 on index 1
TEST-Thr...Thread]  D  Finished removeMediaSourceSync on index 2
EventLogger         D  timeline [eventTime=0.02, mediaPos=0.00, window=0, periodCount=1, windowCount=1, reason=SOURCE_UPDATE
EventLogger         D    period [?]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D  ]
EventLogger         D  mediaItem [eventTime=0.02, mediaPos=0.00, window=0, reason=PLAYLIST_CHANGED]
EventLogger         D  loading [eventTime=0.03, mediaPos=0.00, window=0, period=0, true]
EventLogger         D  timeline [eventTime=0.03, mediaPos=0.00, window=0, period=0, periodCount=2, windowCount=2, reason=SOURCE_UPDATE
EventLogger         D    period [?]
EventLogger         D    period [?]
EventLogger         D    window [?, seekable=false, dynamic=false]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D  ]
EventLogger         D  timeline [eventTime=0.03, mediaPos=0.00, window=0, period=0, periodCount=3, windowCount=3, reason=SOURCE_UPDATE
EventLogger         D    period [?]
EventLogger         D    period [?]
EventLogger         D    period [?]
EventLogger         D    window [?, seekable=false, dynamic=false]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D  ]
EventLogger         D  timeline [eventTime=0.03, mediaPos=0.00, window=0, period=0, periodCount=2, windowCount=2, reason=SOURCE_UPDATE
EventLogger         D    period [?]
EventLogger         D    period [?]
EventLogger         D    window [?, seekable=false, dynamic=false]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D  ]
TrafficStats        D  tagSocket(103) with statsTag=0xffffffff, statsUid=-1
EventLogger         D  audioEnabled [eventTime=0.39, mediaPos=0.00, window=0, period=0]
EventLogger         D  tracks [eventTime=0.39, mediaPos=0.00, window=0, period=0
EventLogger         D    group [
EventLogger         D      [X] Track:0, id=1, mimeType=audio/mp4a-latm, bitrate=128695, codecs=mp4a.40.2, channels=2, sample_rate=48000, language=und, supported=YES
EventLogger         D    ]
EventLogger         D  ]
EventLogger         D  timeline [eventTime=0.39, mediaPos=0.00, window=0, period=0, periodCount=2, windowCount=2, reason=SOURCE_UPDATE
EventLogger         D    period [634.20]
EventLogger         D    period [?]
EventLogger         D    window [634.20, seekable=true, dynamic=false]
EventLogger         D    window [?, seekable=false, dynamic=true]
EventLogger         D  ]
EventLogger         D  loading [eventTime=0.57, mediaPos=0.00, window=0, period=0, false]
```

**Comment by stoyicker** [comment]
> Can you try doing the setup without special threads as a start?

I've tested this by replacing ExoPlayer's control handler by one using the main looper (the handler in ConcatenatingMediaSourceController needs to use a thread of its own because the requirement I'm trying to satisfy is making synchronous additions and removals to the ConcatenatingMediaSource and to achieve that I need a thread that's available to receive the corresponding completion callbacks and release the ExoPlayer control thread). Same result.

Then if the app really needs these additional threads, can you make sure that media sources are added before prepare() is called on the player?

With the same setup as described above but also moving prepare after all calls on the ConcatenatingMediaSource and just before seekToNext, this also doesn't work - in fact there is a significantly smaller amount of logs coming from EventLogger. I would also like to clarify that even if doing this had worked, as far as I can tell it would not be satisfactory as the ConcatenatingMediaSource docs explicitly state that "the list of MediaSources can be modified during playback", meaning that I would expect to need to set at most one MediaSource prior to dispatching prepare(), and then being completely free to add or remove any MediaSources once prepare() has been dispatched.

**Comment by stoyicker** [comment]
I have considered the Playlist API yes, unfortunately that's not an option for me at this time.

Regarding making your PR work with ConcatenatingMediaSource, I have pushed a new commit (to my branch) that moves adding the first child MediaSource to before prepare, as well as some more logging to ensure the order of operations:

```
Finished dispatching all commands
Finished addition on for item id=audio.mp4 index 0
Waiting for addition for item id=audio.mp4 index 0
Finished addMediaSource for item id=audio.mp4 on index 0
playWhenReady changed to true
Received new playback state BUFFERING
Just called prepare
Waiting for addition for item id=dizzy.mp4 index 1
Finished addition on for item id=dizzy.mp4 index 1
Finished addMediaSource for item id=dizzy.mp4 on index 1
Waiting for addition for item id=tears_audio_eng.mp4 index 1
Finished addition on for item id=tears_audio_eng.mp4 index 1
Finished addMediaSource for item id=tears_audio_eng.mp4 on index 1
Waiting for deletion for index 2
Finished deletion for index 2
Finished removeMediaSourceSync on index 2
Received new playback state READY
```

This looks correct to me, yet it still doesn't work correctly because, even though with the change of preparing only after there's one child media source does mean that playback starts, it's only the first item that plays - the seek is completely ignored.

Note that I think there's been a misunderstanding: I don't care which threads things happen on. What I need to do accomplish the synchronicity that's already built into the Playlist API addMediaSource methods, but using ConcatenatingMediaSource's instead, which are asynchronous:

```
// This works fine because ExoPlayerImpl.addMediaSource and removeMediaSource are synchronous
exoPlayer.addMediaSource(firstMediaSource)
exoPlayer.prepare()
exoPlayer.addMediaSource(secondMediaSource)
exoPlayer.addMediaSource(thirdMediaSource)
exoPlayer.removeMediaSource(1)
exoPlayer.seekToNext()
```

```
// This doesn't work (there's no seek) because ConcatenatingMediaSource.addMediaSource and removeMediaSource are asynchronous
concatenatingMediaSource.addMediaSource(firstMediaSource)
exoPlayer.setMediaSource(concatenatingMediaSource)
exoPlayer.prepare()
concatenatingMediaSource.addMediaSource(secondMediaSource)
concatenatingMediaSource.addMediaSource(thirdMediaSource)
concatenatingMediaSource.removeMediaSource(1)
exoPlayer.seekToNext()
```

To make the seekToNext() call work when using ConcatenatingMediaSource I have written ConcatenatingMediaSourceController, which blocks the calling thread and uses another one to run completion callbacks on, with said completion callbacks unlocking the calling thread. If I did everything in the same thread my completion callbacks would never run because the calling thread is locked. So the additional thread in ConcatenatingMediaSourceController is just a means for the goal of making seekToNext() work as if I wasn't using ConcatenatingMediaSource, and the thread created in MainActivity has no purpose other than avoiding the main one. You may replace either of the threads backing the handlers with the main one if you so choose and the results of the tests don't vary.

In conclusion, because even when using ConcantenatingMediaSourceController, which should theoretically make the changes to ConcatenatingMediaSource synchronous, the seek is still not processed correctly, I still believe there is something not working correctly with ConcatenatingMediaSource.
